softlayer/resource_softlayer_file_storage.go

```diff
@@ -0,0 +1,941 @@
+	if err != nil {
+		return fmt.Errorf("Error while creating storage:%s", err)
+	}
+
+	log.Println("[INFO] Creating storage")
+
+	var receipt datatypes.Container_Product_Order_Receipt
+
+	switch storageType {
+	case EnduranceType:
+		receipt, err = services.GetProductOrderService(sess).PlaceOrder(
+			&datatypes.Container_Product_Order_Network_Storage_Enterprise{
+				Container_Product_Order: storageOrderContainer,
+			}, sl.Bool(false))
+	case PerformanceType:
+		receipt, err = services.GetProductOrderService(sess).PlaceOrder(
+			&datatypes.Container_Product_Order_Network_PerformanceStorage_Nfs{
+				Container_Product_Order_Network_PerformanceStorage: datatypes.Container_Product_Order_Network_PerformanceStorage{
+					Container_Product_Order: storageOrderContainer,
+				},
+			}, sl.Bool(false))
+	default:
+		return fmt.Errorf("Error during creation of storage: Invalied storageType %s", storageType)
+	}
+
+	if err != nil {
+		return fmt.Errorf("Error during creation of storage: %s", err)
+	}
+
+	// Find the storage device
+	fileStorage, err := findStorageByOrderId(sess, *receipt.OrderId)
+
+	if err != nil {
+		return fmt.Errorf("Error during creation of storage: %s", err)
+	}
+	d.SetId(fmt.Sprintf("%d", *fileStorage.Id))
+
+	// Wait for storage availability
+	_, err = WaitForStorageAvailable(d, meta)
+
+	if err != nil {
+		return fmt.Errorf(
+			"Error waiting for storage (%s) to become ready: %s", d.Id(), err)
+	}
+
+	// SoftLayer changes the device ID after completion of provisioning. It is necessary to refresh device ID.
+	fileStorage, err = findStorageByOrderId(sess, *receipt.OrderId)
+
+	if err != nil {
+		return fmt.Errorf("Error during creation of storage: %s", err)
+	}
+	d.SetId(fmt.Sprintf("%d", *fileStorage.Id))
+
+	log.Printf("[INFO] Storage ID: %s", d.Id())
+
+	return resourceSoftLayerFileStorageUpdate(d, meta)
+}
+
+func resourceSoftLayerFileStorageRead(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ProviderConfig).SoftLayerSession()
+	storageId, _ := strconv.Atoi(d.Id())
+
+	storage, err := services.GetNetworkStorageService(sess).
+		Id(storageId).
+		Mask(storageDetailMask).
+		GetObject()
+
+	if err != nil {
+		return fmt.Errorf("Error retrieving storage information: %s", err)
+	}
+
+	storageType := strings.Fields(*storage.StorageType.Description)[0]
+
+	// Calculate IOPS
+	iops, err := getIops(storage, storageType)
+	if err != nil {
+		return fmt.Errorf("Error retrieving storage information: %s", err)
+	}
+
+	d.Set("type", storageType)
+	d.Set("capacity", *storage.CapacityGb)
+	d.Set("volumename", *storage.Username)
+	d.Set("hostname", *storage.ServiceResourceBackendIpAddress)
+	d.Set("iops", iops)
+	if storage.SnapshotCapacityGb != nil {
+		snapshotCapacity, _ := strconv.Atoi(*storage.SnapshotCapacityGb)
+		d.Set("snapshot_capacity", snapshotCapacity)
+	}
+
+	// Parse data center short name from ServiceResourceName
```

Comment:
Can you list an example? This might also be a good area to write a unit test for this regex. @renier is there a template for writing unit tests?